### PR TITLE
Add 'new-directory' output specification to item_tasks

### PR DESCRIPTION
**Pull request by sutartmelson**
Client-side fix to #2258

### CODE REVIEW

File: plugins/item_tasks/web_client/models/WidgetModel.js [hunk 1]
```diff
@@ -302,6 +305,8 @@ var WidgetModel = Backbone.Model.extend({
             return 'Value must be a folder';
         } else if (this.get('type') === 'new-file' && type !== 'folder') {
             return 'Value must be a folder';
+        } else if (this.get('type') === 'new-directory' && type !== 'folder') {
```
Comment (brianhelba) [inline plugins/item_tasks/web_client/models/WidgetModel.js]: What if we want to create the new folder at the top level, directly inside a collection (or a user's home directory).

File: plugins/item_tasks/web_client/models/WidgetModel.js [hunk 2]
```diff
@@ -42,6 +42,9 @@ import tinycolor from 'tinycolor2';
  *   * new-file:
  *      an output file (contains an existing folder id and a
  *      name that will be used for the new item.
+ *   * new-directory:
```
Comment (brianhelba) [inline plugins/item_tasks/web_client/models/WidgetModel.js]: Do we really want to call this "directory"? Everything else in Girder uses the terminology of "folders".

Comment (zachmullen) [inline plugins/item_tasks/web_client/models/WidgetModel.js]: +1 for "folder" instead.

File: plugins/item_tasks/web_client/models/WidgetModel.js [hunk 3]
```diff
@@ -302,6 +305,8 @@ var WidgetModel = Backbone.Model.extend({
             return 'Value must be a folder';
         } else if (this.get('type') === 'new-file' && type !== 'folder') {
             return 'Value must be a folder';
+        } else if (this.get('type') === 'new-folder' && !_.contains(['folder', 'collection', 'user'], type)) {
+            return 'Value must be a folder';
```
Comment (brianhelba) [inline plugins/item_tasks/web_client/models/WidgetModel.js]: This error message is not quite accurate any more.

File: plugins/item_tasks/web_client/views/ControlWidget.js [hunk 4]
```diff
@@ -191,14 +194,38 @@ var ControlWidget = View.extend({
...
+        if (type === 'new-folder') {
+            title = 'Create a new folder';
+            help = 'Browse to a path, enter a name, then click "Save"';
+            input = {
+                label: 'Name',
+                placeholder: 'Choose a name for the new folder',
+                validate: (val) => {
+                    // validation on the "new item name"
+                    if (!val) {
+                        return 'Please provide an folder name.';
```
Comment (brianhelba) [inline plugins/item_tasks/web_client/views/ControlWidget.js]: "provide **a** folder name"

File: plugins/item_tasks/web_client/views/ControlWidget.js [hunk 5]
```diff
@@ -191,14 +194,38 @@ var ControlWidget = View.extend({
...
+                    }
+                },
+                default: this.model.get('fileName')
+            };
+            // validation on the parent model
+            validate = (model) => {
+                var type = model.get('_modelType');
+                if (!_.contains(['folder', 'collection', 'user'], type)) {
+                    return 'Invalid parent type, please choose a folder.';
```
Comment (brianhelba) [inline plugins/item_tasks/web_client/views/ControlWidget.js]: Not quite accurate.

File: plugins/item_tasks/web_client/views/ControlWidget.js [hunk 6]
```diff
@@ -191,14 +194,38 @@ var ControlWidget = View.extend({
...
+            input = {
+                label: 'Name',
+                placeholder: 'Choose a name for the new folder',
+                validate: (val) => {
+                    // validation on the "new item name"
+                    if (!val) {
+                        return 'Please provide an folder name.';
+                    }
+                },
+                default: this.model.get('fileName')
```
Comment (brianhelba) [inline plugins/item_tasks/web_client/views/ControlWidget.js]: I'm not sure: do we expect this value to be set from the task spec?

**Comment by sutartmelson** [comment]
@brianhelba My initial thought was to wait for the server side to add tests.. but I now realize I need not wait for that. I'll add tests today.

File: plugins/item_tasks/web_client/views/ControlWidget.js (hunk 6 above)
Comment (sutartmelson) [inline plugins/item_tasks/web_client/views/ControlWidget.js]: I'm not sure. Is this task spec outlined anywhere? I tried looking at girder_worker but didn't find comprehensive format specifications.

Comment (jbeezley) [inline plugins/item_tasks/web_client/views/ControlWidget.js]: The spec has largely grown organically.  The most comprehensive documentation is here: https://github.com/girder/girder/blob/master/plugins/item_tasks/web_client/models/WidgetModel.js#L5

**Comment by zachmullen** [comment]
I just tested this out, and it's working well. One thing that isn't working yet is populating the widget based on a previous job, which is done [here](https://github.com/girder/girder/blob/master/plugins/item_tasks/web_client/views/TaskRunView.js#L61-L80).

File: plugins/item_tasks/web_client/collections/WidgetCollection.js [hunk 7]
```diff
@@ -22,6 +22,10 @@ var WidgetCollection = Backbone.Collection.extend({
                     params[m.id + '_girderFolderId'] = m.value().get('folderId');
                     params[m.id + '_name'] = m.value().get('name');
                     break;
+                case 'new-folder':
```
Comment (sutartmelson) [inline plugins/item_tasks/web_client/collections/WidgetCollection.js]: I more or less blindly edited this file, so I'm not sure if this is correct or where this is actually used.

**Comment by zachmullen** [comment]
Hm, that last commit isn't working for me, the name is not populated from the previous job in the directory output field.

**Comment by sutartmelson** [comment]
@zachmullen  Apparently it's `parent_type` and not `parentType`. I don't really understand how I got it to work earlier then if that's wrong. Regardless, it should work now.

**Comment by zachmullen** [comment]
🎉  this is now working on the client side. One last thing I'd like to fix on this branch: the server side still complains with `Error: Invalid output parent type: user.`. Could we fix the validation to allow parent_type to be user or collection in the case of a new-folder output? This would be in `_transformOutputs` I think.

File: plugins/item_tasks/server/rest.py [hunk 8]
```diff
@@ -168,6 +168,29 @@ def _transformOutputs(self, outputs, token, job):
...
+    def _validateOutputParentType(self, outputId, parentType, outputSpec):
+        """
+        Checks if the output parent type is compatible with the output type.
+        """
+
+        # Find the corresponding output specification for the given outputID
+        output = None
+        for out in outputSpec:
+            if outputId == out['id']:
+                output = out
```
Comment (zachmullen) [inline plugins/item_tasks/server/rest.py]: and `break`

File: plugins/item_tasks/server/rest.py [hunk 9]
```diff
@@ -168,6 +168,29 @@ def _transformOutputs(self, outputs, token, job):
...
+        if output:
+            if output['type'] == 'new-file' and parentType not in {'item', 'folder'}:
+                return False
+            elif output['type'] == 'new-folder' and\
+                    parentType not in {'folder', 'user', 'collection'}:
+                return False
+            else:
+                return True
+        else:
+            return False
```
Comment (zachmullen) [inline plugins/item_tasks/server/rest.py]: This should probably raise its own `ValidationException` since it's not really an invalid parent type, but actually an invalid output ID.

**Comment by zachmullen** [comment]
This works for me now, thanks!

@brianhelba could you give a final review on this?

File: plugins/item_tasks/server/rest.py [hunk 10]
```diff
@@ -168,6 +168,30 @@ def _transformOutputs(self, outputs, token, job):
 
         return transformed
 
+    def _validateOutputParentType(self, outputId, parentType, outputSpec):
+        """
+        Checks if the output parent type is compatible with the output type.
+        """
+
+        # Find the corresponding output specification for the given outputID
+        output = None
```
Comment (brianhelba) [inline plugins/item_tasks/server/rest.py]: Python loop iteration variables persist in the scope outside, so you could use `out` in place of `output` below (or vice versa)

File: plugins/item_tasks/server/rest.py [hunk 11]
```diff
@@ -168,6 +168,30 @@ def _transformOutputs(self, outputs, token, job):
...
+        for out in outputSpec:
+            if outputId == out['id']:
+                output = out
+                break
+
+        # If a corresponding output is found, check if its parent type is valid
+        if output:
```
Comment (brianhelba) [inline plugins/item_tasks/server/rest.py]: Python `for` loops take an optional `else:` clause, which only runs if the loop iterates to completion without any `break` being triggered. You could use this to eliminate both sides of the `if output:` conditional, moving the contained logic up one level.

Comment (mgrauer) [inline plugins/item_tasks/server/rest.py]: The for/else in Python is frequently misunderstood.  I'd avoid using it if possible.

https://aroberge.blogspot.com/2015/04/loopy-else-in-python.html

Comment (zachmullen) [inline plugins/item_tasks/server/rest.py]: I actually really like the for/else language feature, but agree that it is confusing because it's unfamiliar since it's not present in other programming languages. That said, it's a language construct that exists precisely for these sorts of cases, and it's decidedly more elegant than the more C-like construct that is currently written.

I agree with the article that the main source of confusion is the term `else` as opposed to a more explicit one like `nobreak`.

TLDR: While I personally would have written this using for/else, I do not feel that it needs to be changed since it works as written since there are tradeoffs.

File: plugins/item_tasks/server/rest.py [hunk 12]
```diff
@@ -168,6 +168,30 @@ def _transformOutputs(self, outputs, token, job):
...
+            if output['type'] == 'new-file' and parentType not in {'item', 'folder'}:
+                return False
+            elif output['type'] == 'new-folder' and\
+                    parentType not in {'folder', 'user', 'collection'}:
+                return False
+            else:
+                return True
+        else:
+            raise ValidationException('Invlaid outpud id: %s.' % outputId)
```
Comment (mgrauer) [inline plugins/item_tasks/server/rest.py]: Invalid output
